test_builtins.py

```diff
@@ -0,0 +1,130 @@
+
+    def test_builtins(self) -> None:
+        env: Dict[SSym, Value] = {}
+        runner.add_intrinsics(env)
+        runner.add_builtins(env)
+
+        with self.assertRaises(errors.Trap, msg="(trap)"):
+            run(env, '(trap)')
+        with self.assertRaises(errors.Trap, msg="(trap)"):
+            run(env, '(assert false)')
+        self.assertEqual(run(env, '(typeof 42)'), SSym('number'))
+        self.assertEqual(run(env, '(typeof [1 2 3])'), SSym('vector'))
+        self.assertEqual(run(env, "(typeof 'hi)"), SSym('symbol'))
+        self.assertEqual(run(env, '(not true)'), SBool(False))
+        self.assertEqual(run(env, '(not false)'), SBool(True))
+        with self.assertRaises(errors.Trap, msg="(trap)"):
+            run(env, '(not 42)')
+
+        self.assertEqual(run(env, '(number? 42)'), SBool(True))
+        self.assertEqual(run(env, '(number? [])'), SBool(False))
+        self.assertEqual(run(env, "(symbol? 'hi)"), SBool(True))
+        self.assertEqual(run(env, '(symbol? 42)'), SBool(False))
+        self.assertEqual(run(env, '(vector? [])'), SBool(True))
+        self.assertEqual(run(env, '(vector? 42)'), SBool(False))
+        self.assertEqual(run(env, '(function? (lambda () []))'), SBool(True))
+        self.assertEqual(run(env, '(function? 42)'), SBool(False))
+        self.assertEqual(run(env, '(bool? true)'), SBool(True))
+        self.assertEqual(run(env, '(bool? false)'), SBool(True))
+        self.assertEqual(run(env, '(bool? 42)'), SBool(False))
+        self.assertEqual(run(env, '(pair? 42)'), SBool(False))
+        self.assertEqual(run(env, '(pair? [])'), SBool(False))
+        self.assertEqual(run(env, '(pair? [1])'), SBool(False))
+        self.assertEqual(run(env, '(pair? [1 2])'), SBool(True))
+        self.assertEqual(run(env, '(nil? [1 2])'), SBool(False))
+        self.assertEqual(run(env, '(nil? [])'), SBool(True))
+
+        self.assertEqual(run(env, '(+ 39 13)'), SNum(39 + 13))
+        self.assertEqual(run(env, '(- 39 13)'), SNum(39 - 13))
+        self.assertEqual(run(env, '(* 39 13)'), SNum(39 * 13))
+        self.assertEqual(run(env, '(/ 39 13)'), SNum(39 // 13))
+        self.assertEqual(run(env, '(% 39 13)'), SNum(39 % 13))
+
+        for op in '+-*/%':
+            with self.assertRaises(errors.Trap, msg="(trap)"):
+                run(env, f'({op} 39 [])')
+            with self.assertRaises(errors.Trap, msg="(trap)"):
+                run(env, f'({op} [] 13)')
+
+        with self.assertRaises(errors.Trap, msg="(trap)"):
+            run(env, '(/ 1 0)')
+        with self.assertRaises(errors.Trap, msg="(trap)"):
+            run(env, '(% 1 0)')
+
+        with self.assertRaises(errors.Trap, msg="(trap)"):
+            run(env, '(symbol= 1 0)')
+        self.assertEqual(run(env, "(symbol= 'a 'b)"), SBool(False))
+        self.assertEqual(run(env, "(symbol= 'a 'a)"), SBool(True))
+        # I don't want to test the other foo= functions, they're very similar
```

Comment:
I got you a present:
```
        # This first one has an uncaught exception because there aren't any type checks for pointer=. You can decide what to do about that.
        with self.assertRaises(errors.Trap, msg="(trap)"):
            run(env, "(pointer= 'a false)")
        self.assertEqual(run(env, "(pointer= (lambda () 0) (lambda () 0))"),
                         SBool(False))
        self.assertEqual(run(env, "(define (func) 42) (pointer= func func)"),
                         SBool(True))
        self.assertEqual(run(env, "(pointer= 0 0)"), SBool(True))

        with self.assertRaises(errors.Trap, msg="(trap)"):
            run(env, "(number= 'num 1)")
        self.assertEqual(run(env, "(number= 42 43)"), SBool(False))
        self.assertEqual(run(env, "(number= 42 42)"), SBool(True))

        with self.assertRaises(errors.Trap, msg="(trap)"):
            run(env, "(number< 'num 1)")
        self.assertEqual(run(env, "(number< 42 42)"), SBool(False))
        self.assertEqual(run(env, "(number< 42 43)"), SBool(True))
```